### PR TITLE
Fixed dtype of a converted unit

### PR DESCRIPTION
**Pull request by dkillick**
Fixes two dtype issues when converting an Iris unit:
 * the datatype specified by the `ctype` kwarg to `iris.unit.Unit.convert` is not being enforced, and
 * big-endian arrays are not being handled correctly.

Both of these issues are fixed by enforcing the datatype specified by the `ctype` kwarg. This caused a number of knock-on test failures, all of which boiled down to the datatype of the converted unit being implied rather than fixed.

To appropriately fix the datatype the `ctype` kwarg is now explicitly passed through `cube.convert_units` and `coord.convert_units`.

### CODE REVIEW

**Comment by rhattersley** [comment]
NB. The docs for `Unit.convert` state, "Conversion is done *in-place* for numpy arrays." This is (mostly*) not true and could do with fixing in a separate PR.

*) It's only true when the source and target units are equal, in which case no computation is necessary and the original array is returned.

File: lib/iris/unit.py [hunk 1]
```diff
@@ -1869,18 +1872,15 @@ def convert(self, value, other, ctype=FLOAT64):
...
                     if isinstance(value_copy, np.ndarray):
-                        # Can only handle array of np.float32 or np.float64 so
-                        # cast array of ints to array of floats of requested
+                        # Can only handle dtypes of np.float32 or np.float64 so
+                        # cast `value_copy` to floats of requested
                         # precision.
-                        if issubclass(value_copy.dtype.type, np.integer):
-                            value_copy = value_copy.astype(
-                                _ctypes2numpy[ctype])
+                        value_copy = value_copy.astype(_ctypes2numpy[ctype])
```
Comment (pelson) [inline lib/iris/unit.py]: Proposal: revert this change, byteswap if necessary, and improve the ctype keyword documentation to mention the fact that it is only for integer types.

Comment (rhattersley) [inline lib/iris/unit.py]: > it is only for integer types.

It's also used for non-array values.

**Comment by rhattersley** [comment]
What's the real driver behind this PR? Is it the big-endian issue?

**Comment by dkillick** [comment]
@rhattersley primarily this PR is to solve the endian-ness issue. Associated with this it was pointed out that the `ctype` kwarg was usually ignored. In fixing the second issue I found I had also fixed the first issue.

File: lib/iris/unit.py (hunk 1 above)
Comment (rhattersley) [inline lib/iris/unit.py]: > revert this change, byteswap if necessary, and improve the ctype keyword documentation

@dkillick & I had a chat offline - he now plans to take the approach suggested above.

**Comment by dkillick** [comment]
This latest commit switches the approach taken by this PR from the one originally proposed to the opposite proposal as championed by @rhattersley and @pelson. It has noticeably simplified the PR and caused far fewer incidental test failures and other code knock-on impacts.

File: lib/iris/unit.py [hunk 2]
```diff
@@ -1814,8 +1814,9 @@ def convert(self, value, other, ctype=FLOAT64):
             Target unit to convert to.
         * ctype (ctypes.c_float/ctypes.c_double):
             Floating point 32-bit single-precision (iris.unit.FLOAT32) or
-            64-bit double-precision (iris.unit.FLOAT64) of conversion. The
-            default is 64-bit double-precision conversion.
+            64-bit double-precision (iris.unit.FLOAT64) used for conversion
+            when `value` is of integer type. The default is 64-bit
```
Comment (pelson) [inline lib/iris/unit.py]: More like when the type is not float?

File: lib/iris/unit.py [hunk 3]
```diff
@@ -1875,6 +1876,11 @@ def convert(self, value, other, ctype=FLOAT64):
                         if issubclass(value_copy.dtype.type, np.integer):
                             value_copy = value_copy.astype(
                                 _ctypes2numpy[ctype])
+                        # Convert big-endian arrays because udunits doesn't
```
Comment (pelson) [inline lib/iris/unit.py]: We don't know that this is true for big-endian systems. Alas, I don't propose we check that - the tests will catch the problem if it comes up :+1:

**Comment by dkillick** [comment]
I've made changes based on @pelson's comments:
 * the method docstring now makes it clear that `ctype` will be used to convert any input `value` that is not of type float, and
 * any input `value` with explicit endianness will be converted to having native endianness. This generalises behaviour for both big- and little-endian systems, and means the comment describing the behaviour is more accurate.

**Comment by rhattersley** [comment]
> the method docstring now makes it clear that `ctype` will be used to convert any input `value` that is not of type float

I know @pelson said, "More like when the type is not float?" but I don't think that's right. It gets used for integer arrays(*) and ___any___ scalar. For example, `src_unit.convert(np.float32(2.4), 'm')` will still force a 64-bit conversion.

*) Admittedly, because the array handling raises an error for unrecognised dtypes, saying "integer arrays" is pretty much equivalent to saying "not floating point arrays" but that's not the real issue.

**Comment by dkillick** [comment]
I've had another go at wording the description for the `ctype` argument in the `unit.convert` docstring. It now literally states the only time that the value of the `ctype` argument will be used.

**Comment by pelson** [comment]
:+1: from me

**Comment by dkillick** [comment]
> now literally states

It may have been literal but it was not complete... hopefully this fixes it! @rhattersley?

**Comment by rhattersley** [comment]
:+1:

**Comment by rhattersley** [comment]
It all boiled down to quite a small change in the end - any chance of a squash?

**Comment by dkillick** [comment]
> any chance of a squash?

With the proviso that I keep the original commit separate, yes. I think it would be good not to lose the history of the other approach.

**Comment by dkillick** [comment]
If we can just hold our horses :horse: for a second, I need to rebuild my history :cherries: ... Watch this space!

**Comment by dkillick** [comment]
Phew, that's much better :sweat_smile:

**Comment by rhattersley** [comment]
Thanks :smile:
